### ISSUE TITLE
L16_Convolutional_Neural_Network.ipynb

### ISSUE THREAD

**Issue by uzinny04** [uzinny04/AI_Appl#4]
#### 다음과 같은 CNN 모델을 작성해보자
- Input
    - Input type: torch.Tensor
    - Input shape: (?, 1, 28, 28)
      - 여러장의, 흑백, 28x28 size의 이미지라고 가정하자
- Layers
  - Layer 1
  - Conv2d >> C: 32, Kernel size (필터 크기): 3, Stride: 1, Padding: 1
  - ReLU
  - MaxPool >> Kernel size: 2, Stride: 2
  - 입-출력 (?, 1, 28, 28) >> (?, 32, 14, 14)
- Layer 2
  - Conv2d >> C: 64, Kernel size (필터 크기): 3, Stride: 1, Padding: 1
  - ReLU
  - MaxPool >> Kernel size: 2, Stride: 2
  - 입-출력 (?, 32, 14, 14) >> (?, 64, 7, 7)
- Layer 3
  - Conv2d >> C: 128, Kernel size (필터 크기): 3, Stride: 1, Padding: 1
  - ReLU
  - MaxPool >> Kernel size: 2, Stride: 2, Padding: 1
  - 입-출력 (?, 64, 7, 7) >> (?, 128, 4, 4)
- Layer 4
  - Linear >> input: 4x4x128 output: 625
  - ReLU
  - Dropout
  - 입-출력 (4x4x128) >> (625)
- Layer 5
  - Linear >> input: 625 output: 10
  - Softmax (pytorch의 Cross Entropy Loss 함수를 사용하는 것을 감안한다)

# import
import torch  # torch import
import torchvision.datasets as dsets # torchvision의 데이터셋 import
import torchvision.transforms as transforms # 이미지 변환 기능 import
import torch.nn.init # tensor에 초깃값 주는 함수 import

device = 'cuda' if torch.cuda.is_available() else 'cpu' # GPU 사용 가능하면 device 값이 cuda, 아니면 cpu

torch.manual_seed(777)  # 랜덤 시드 고정
if device == 'cuda':    # GPU 사용 가능일 경우 랜덤 시드 고정
    torch.cuda.manual_seed_all(777)

# 클래스로 CNN model 설계
class CNN(torch.nn.Module):

    def __init__(self):
        super(CNN, self).__init__()
        # N = Batch Size
        #    l1 Input = (N, 1, 28, 28)
        #    Conv 결과 = (N, 32, 28, 28)
        #    Pool 결과 = (N, 32, 14, 14) : 사이즈만 1/4 로 줄어듦 (가로 절반, 세로 절반)
        self.layer1 = torch.nn.Sequential(   # Sequential : 밑 3줄(3개의 layer) 한번에 처리 (가독성을 높여줌)
            torch.nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),  # 3,1,1로 설정할경우 입출력 사이즈 같음
            torch.nn.ReLU(),   # 활성화 함수 Relu 사용
            torch.nn.MaxPool2d(kernel_size=2, stride=2))
         
        #    l2 Input = (N, 32, 14, 14)
        #    Conv 결과 = (N, 64, 14, 14)
        #    Pool 결과 = (N, 64, 7, 7)
        self.layer2 = torch.nn.Sequential( 
            torch.nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(kernel_size=2, stride=2))
        
        #    l3 Input = (N, 64, 7, 7)
        #    Conv 결과 = (N, 128, 7, 7)
        #    Pool 결과 = (N, 128, 4, 4)
        self.layer3 = torch.nn.Sequential(
            torch.nn.Conv2d(64, 128, kernel_size=3, stride=1, padding=1),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(kernel_size=2, stride=2,padding=1)) # 패딩 1 추가
       
        #    l4 Input = (N, 128, 4, 4) >> output = 625
        self.layer4= torch.nn.Sequential(
            torch.nn.Linear(4*4*128, 625, bias=True),
            torch.nn.ReLU(),
            torch.nn.Dropout(0.5) #설정한 노드중 50% 끄고 계산한다 : 과적합 방지

        )
        torch.nn.init.xavier_uniform_(self.layer4.weight) # layer4 (전결합층) 만 가중치 초기화
        # Dropout 의 경우, 과적합 방지용이라고 생각하면 된다. 은닉층에 많은 노드를 전부 계산하게 되면 과적합 가능성이 올라가기 때문에, 
        # 어느 정도를 끄고 돌리면 어떤 결과가 나오는지 계산을 해서 임의로 노드를 끄는 것이다. 이 어느 정도에 해당하는 값이 이번에는 0.25가 된다. 즉, 설정한 노드 중 25퍼센트는 끄면서 계산을 하는 것이다.
        # L5 Linear 625 inputs -> 10 outputs
        self.layer5 = torch.nn.Linear(625, 10, bias=True)
        torch.nn.init.xavier_uniform_(self.layer5.weight) # xavier : 무작위가 아닌 방법으로 weight를 초기화 
        #기존에는 weight를 무작위로 초기화 했는데 무작위가 아닌 방법으로 weight를 초기화해서 성능을 높인다.그 방법 중 하나인 Xavier

    def forward(self, x):
        out = self.layer1(x)  # 전 층의 출력값이 다음 층의 입력값으로 들어감.
        out = self.layer2(out)
        out = self.layer3(out)
        out = out.view(out.size(0), -1) # 첫번째 차원(Batch) 제외한 차원 펼치기
        out = self.layer4(out)
        out = self.layer5(out)   
        
        return out
